### ISSUE TITLE
Only some ligatures are being applied

### ISSUE THREAD

**Issue by wezm** [zed-industries/zed#12176]
### Check for existing issues

- [X] Completed

Possibly related to https://github.com/zed-industries/zed/issues/11127 but that is specifically about the terminal.

### Describe the bug / provide steps to reproduce it

I have enabled ligatures in my `settings.json`:

```json
  "buffer_font_family": "PragmataPro Mono Liga",
  "buffer_font_features": {
    "calt": true
  },
```

However only some of the expected ligatures are being applied. For example if I type `<!--` it ligates correctly, as does `=>` but `!=` should ligate to `≠` and `->` to something like `→` but they do not.

These are documented in [the font handbook](https://fsd.it/pragmatapro/Handbook.png) as part of `calt` as well as shown in other editors.

### Environment

Zed: v1.0.0 (Zed Dev 71a94c775b752af59b9be7f41e78afef5e63615e)
OS: Linux 1.0.0 (X11)
Memory: 62.5 GiB
Architecture: x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

These screenshots show the difference in rendering between Zed and JetBrains Rust Rover both using the same font with ligatures enabled.

Zed:

In this one `=>` _is_ ligating but  `->`, `>>` in `read::<CmapSubtable<'_>>()` and `//` are not.

<img width="590" alt="Screenshot from 2024-05-23 11-00-28" src="https://github.com/zed-industries/zed/assets/21787/03e14d22-2cb8-4568-8e22-4cef994becc5">

Rust Rover:

<img width="581" alt="Screenshot from 2024-05-23 10-58-04" src="https://github.com/zed-industries/zed/assets/21787/285f7181-4a7b-4b0e-93f0-bc65746107fc">

Zed:

In this one it appears that `[[` and `]]` are ligating but `!=` is not.

<img width="173" alt="Screenshot from 2024-05-23 11-00-01" src="https://github.com/zed-industries/zed/assets/21787/48d79a7a-8923-4e69-8c5d-e226f8d532e2">

Rust Rover:

<img width="216" alt="Screenshot from 2024-05-23 10-59-52" src="https://github.com/zed-industries/zed/assets/21787/ad49a909-c12b-4058-947a-26b80947c22d">

**Comment by JunkuiZhang** [comment]
Linux implementation dose not support liga for now.

**Comment by wezm** [comment]
> Linux implementation dose not support font features for now.

Ahh I see, well I guess this can act as a tracking issue for the implementation. Here's the TODO:

https://github.com/zed-industries/zed/blob/3ec94697b4c3def47209674cf08665fd1674b8ad/crates/gpui/src/platform/cosmic_text/text_system.rs#L198-L204

**Comment by peppidesu** [comment]
Seems like we need to wait for https://github.com/pop-os/cosmic-text/issues/229 to be resolved

**Comment by veryshyjelly** [comment]
Same thing happening in windows

**Comment by JunkuiZhang** [comment]
> Same thing happening in windows

It's a bug introduced by #13854 , and the changes have been reverted via #14045 .

**Comment by flyInRust** [comment]
Same thing happening in ubuntu 24.04 zed 0.143.6

**Comment by flyInRust** [comment]
Same thing happening in ubuntu 24.04 zed 0.143.6

**Comment by t9t** [comment]
I'm facing this issue as well on Arch Linux using Zed 0.144.3.

@JunkuiZhang and @peppidesu could you explain in short how you came to the conclusion that this is about missing font feature support?

I am asking because some ligatures work fine, while others do not. E.g. as @wezm noted (and as you can see in their screenshot), `=>` works while `!=` does not. Some more examples of working ligatures are `==`, `:=`, `<=`, and `>=`. Not working are e.g. `->`, `-->`.

`calt` is a default font feature, and if I understand https://github.com/zed-industries/zed-fonts/blob/main/params/ligation-set.toml correctly, it should include the not-working `!=` by default (through `exeq`; i.e. not dependent on font features). Though I may be misunderstanding how default font features work.

![image](https://github.com/user-attachments/assets/89f8555e-2e8f-43a0-89bf-c7e85a88cc31)

**Comment by JunkuiZhang** [comment]
@t9t I implemented font feature support on Windows and macOS. However, it is currently impossible to implement font features on Linux because it uses the cosmic text system, which does not support setting font features. See https://github.com/pop-os/cosmic-text/issues/229

**Comment by JunkuiZhang** [comment]
Linux uses `CosmicTextSystem`:

https://github.com/zed-industries/zed/blob/7d30175527b0a588e6414f6bb562fb9d25c3c9ef/crates/gpui/src/platform/linux/platform.rs#L107

You can see that `FontFeatures` is not used:

https://github.com/zed-industries/zed/blob/7d30175527b0a588e6414f6bb562fb9d25c3c9ef/crates/gpui/src/platform/cosmic_text/text_system.rs#L202-L208

**Comment by t9t** [comment]
Thanks a lot. Let me rephrase my question. I would like to understand how the lack of implementation of font features lead to some ligatures missing, while other ligatures work.

**Comment by JunkuiZhang** [comment]
I guess this is some kind of default behavior that cosmic text provides? 

Sorry I know little about cosmic text.

**Comment by andre0xFF** [comment]
Also not seeing ligatures on Ubuntu 22.04.4 LTS zed 0.144.4

**Comment by ruskiiamov** [comment]
same issue
ubuntu 22.04.4
zed 0.147.2

**Comment by cacilhas** [comment]
I’m on Arch Linux, using Zed 0.162.3 with Fira Code, and I got the same issue.

This is how it looks in Zed:
![Zed](https://github.com/user-attachments/assets/72b8966b-4f64-4930-b4f9-1380a340cdab)

And this is how it looks in Neovim over Kitty with the same font:
![Neovim+Kitty](https://github.com/user-attachments/assets/9164625d-f6e7-4fdf-9de7-4d1339bd7f7b)

**Comment by peppidesu** [comment]
Hi, I was still able to reproduce this last week

**Comment by zapazite** [comment]
I think someone need to make the same pull request as [#8928](https://github.com/zed-industries/zed/issues/8928) but for **Linux** version of **Zed**. I am not sure so It is just an Idea, I am not an expert.

**Comment by peppidesu** [comment]
> I think someone need to make the same pull request as [#8928](https://github.com/zed-industries/zed/issues/8928) but for **Linux** version of **Zed**. I am not sure so It is just an Idea, I am not an expert.
@zapazite The problem lies in the fact that font rendering is OS specific, and changes are required upstream to properly resolve the issue without outrageous hacks.

**Comment by peppidesu** [comment]
Quick update: https://github.com/pop-os/cosmic-text/issues/229 was fixed yesterday, and I implemented font features for linux in #27808. Sadly, it looks like this does not fix this issue, which means I no longer have any idea what might be causing this.

**Comment by COca-CO-la** [comment]
~~But it was properly rendered using calt in `Agent Menu`'s Threads. Only not working in buffer.~~
**It only have partial support with a few working ligature.**

**Comment by Igonato** [comment]
@COca-CO-la where exactly? Can you post a screenshot? For me they are broken everywhere:

![Image](https://github.com/user-attachments/assets/c50b7374-0fab-4aab-b05f-9c18421d7d54)

**Comment by ikarius** [comment]
Not as complete as previous comment (thx), but definitely partial support, not complete :

![Image](https://github.com/user-attachments/assets/b0416f42-70f0-48a1-abcf-b1b2dabb0169)

Update : 

![Image](https://github.com/user-attachments/assets/38a5b205-acc2-4e5b-a195-2e92ced11cfb)

**Comment by Igonato** [comment]
Oh, if you want to copy&paste it, here it is:

```js
    // Ligatures check:
    // -<< -< -<- <-- <--- <<- <- -> ->> --> ---> ->- >- >>-
    // =<< =< =<= <== <=== <<= <= => =>> ==> ===> =>= >= >>=
    // <-> <--> <---> <----> <=> <==> <===> <====> :: ::: __
    // <~~ </ </> /> ~~> == != /= ~= <> === !== !=== =/= =!=
    // <: := *= *+ <* <*> *> <| <|> |> <. <.> .> +* =* =: :>
    // (* *) /* */ [| |] {| |} ++ +++ \/ /\ |- -| <!-- <!---
```

I got it from [Iosevka homepage](https://typeof.net/Iosevka/) btw.

**Comment by COca-CO-la** [comment]
> [@COca-CO-la](https://github.com/COca-CO-la) where exactly? Can you post a screenshot? For me they are broken everywhere:
> 
> ![Image](https://github.com/user-attachments/assets/c50b7374-0fab-4aab-b05f-9c18421d7d54)

![Image](https://github.com/user-attachments/assets/40eaa4c8-ba67-4203-a91b-b01b504af53f)
~~IDK why this happened.~~ **~~Though~~ Yes, the ligatures test you provide didn't work here either, it doesn't show nerd ligatures.**

Update: it seems like not a ligatures but just a font weight problem. This look difference is not caused by a properly working ligatures.
**The ligatures on Linux remains malfunction.**

**Comment by laurentpayot** [comment]
Same issue on Ubuntu 25.04 with JetBrains Mono:

```
-- --- == === != !== =!= =:= =/= <= >= && &&& &= ++ +++ *** ;; !! ?? ??? ?: ?. ?= <: :< :> >: <:<
<> <<< >>> << >> || -| _|_ |- ||- |= ||= ## ### #### #{ #[ ]# #( #? #_ #_( #: #! #= ^= <$> <$ $>
<+> <+ +> <*> <* *> </ </> /> <!-- <#-- --> -> ->> <<- <- <=< =<< <<= <== <=> <==> ==> => =>> >=>
>>= >>- >- -< -<< >-> <-< <-| <=| |=> |-> <-> <~~ <~ <~> ~~ ~~> ~> ~- -~ ~@ [||] |] [| |} {| [< >]
|> <| ||> <|| |||> <||| <|> ... .. .= ..< .? :: ::: := ::= :? :?> // /// /* */ /= //= /== @_ __
??? <:< ;;;
```
becomes

<img width="958" height="167" alt="Image" src="https://github.com/user-attachments/assets/9083d65b-f4b2-4f26-8e81-8c5c9bca745c" />

For instance, notice the arrow `->`, quite annoying for functional programming languages.

**Comment by Igonato** [comment]
Linking this PR from pop-os/cosmic-text for visibility:

https://github.com/pop-os/cosmic-text/pull/392

~~TL;DR: It fixes the issue, but possibly introduces more in the process, so it isn't merged yet.~~
TL;DR: It "fixes" the issue (in a non-generic way), but introduces more in the process, so its not going to get merged yet.

**Comment by peppidesu** [comment]
> TL;DR: It fixes the issue

@Igonato This PR does not in fact fix the issue. It "fixes" it for one particular font by hard-coding the ligatures of that font, then uses it to remove candid breakpoints that intersect them. This is wrong for 2 reasons:
1. Obviously, this won't work for any other font than the one chosen.
2. Even if a breakpoint intersects a ligature, the text should still be broken when necessary. Not doing this would go against the Unicode specification.

**Comment by hickscorp** [comment]
I'm on Zed `0.198.2` and I can confirm that this is broken on **some** ligatures.
Using Fira Code, notice how the ones in the first line doesn't render, while the ones in second line renders:

<img width="199" height="144" alt="Image" src="https://github.com/user-attachments/assets/1875eca5-db2b-4bc5-9ec7-629ffe1f9113" />

**Comment by wezm** [comment]
Still an issue.

**Comment by hickscorp** [comment]
Still an issue for me as well on `0.212.7`.

<img width="227" height="119" alt="Image" src="https://github.com/user-attachments/assets/a8ac1a5b-d88d-48cb-8c29-6b14512a1fd4" />

My apologies for attaching an image of code - it seems necessary in this case ;)

**Comment by ikarius** [comment]
As for today, version `0.212.7`:

<img width="977" height="309" alt="Image" src="https://github.com/user-attachments/assets/98e4ac8a-07a7-4307-ae29-10a5d3001305" />

**Comment by miguelraz** [comment]
Sorry all for the noise, the bot got an issue upstream and went haywire: https://github.com/actions/stale/issues/1302.

**Comment by SeaRoll** [comment]
This happens on Fedora with Fira code with same as above, only some works

**Comment by Somnia1337** [comment]
Still an issue on Linux, Zed 0.213.7 with .ZedMono

**Comment by andradenathan** [comment]
it's still an issue on Ubuntu using .ZedMono or FiraCode or JetBrains Mono

**Comment by jackpot51** [comment]
It does on Linux

**Comment by adb-sh** [comment]
@laurentpayot  i think this means that some characters are treated as an opportunity to break to a new line, like space. ligatures seem to be applied after parsing possible line breaks, which puts these special characters out of scope to render the ligatures

**Comment by laurentpayot** [comment]
Oh I see. "opportunities" is the key word.

**Comment by hickscorp** [comment]
For me (Fedora, KDE latest, zed latest), it doesn't change anything. The examples I provided screenshots for, and the examples other have provided still are broken.

**Comment by ikarius** [comment]
Still the same for me (Fedora Silverblue). Still hooked to Neovim / Helix. But one day, I will give it a try :)

**Comment by adb-sh** [comment]
<img width="700" height="466" alt="Image" src="https://github.com/user-attachments/assets/2fc47a43-58e4-449f-9056-80741aa660c6" />

looks good to me :D
